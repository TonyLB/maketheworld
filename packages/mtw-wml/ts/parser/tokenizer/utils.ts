import { TokenBase, Tokenizer, TokenError, isTokenError } from './baseClasses'
import SourceStream from './sourceStream'

type CheckSubTokenizerProps<T extends TokenBase & { type: string }> = {
    currentBuffer: string;
    startIdx: number;
    subTokenizer: Tokenizer<T>,
    sourceStream: SourceStream
}

type CheckSubTokenizerReturn = {
    returnBuffer?: string;
    error?: TokenError;
}

export const checkSubTokenizer = <T extends TokenBase & { type: string }>({
    currentBuffer, startIdx, subTokenizer, sourceStream
}: CheckSubTokenizerProps<T>): CheckSubTokenizerReturn => {
    const checkSubTokens = subTokenizer(sourceStream)
    if (checkSubTokens) {
        if (isTokenError(checkSubTokens)) {
            return {
                error: {
                    type: 'Error',
                    source: currentBuffer + checkSubTokens.source,
                    startIdx,
                    endIdx: checkSubTokens.endIdx,
                    message: checkSubTokens.message
                }
            }
        }
        else {
            return {
                returnBuffer: currentBuffer + checkSubTokens.source
            }
        }
    }
    else {
        return {}
    }
}
