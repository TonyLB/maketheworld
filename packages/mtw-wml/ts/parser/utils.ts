import { ParseStackTagOpenEntry, ParseError, ParseTag } from "./baseClasses"

type ValidatePropertiesValueType = 'boolean' | 'key' | 'expression' | 'literal'

type ValidatePropertiesItem = Record<string, ValidatePropertiesValueType[]>

type ValidatePropertiesProps = {
    open: ParseStackTagOpenEntry;
    required: ValidatePropertiesItem;
    optional?: ValidatePropertiesItem;
}

export const validateProperties = <T extends Record<string, string | boolean>>({ open, required, optional = {} }: ValidatePropertiesProps): T | ParseError => {
    const illegalEntry = Object.entries(open.properties).reduce<ParseError | undefined>((previous, [key, value]) => {
        if (previous) {
            return previous
        }
        if (!(key in required || key in optional)) {
            return {
                tag: 'Error',
                message: `'${key}' is not a legal property on ${open.tag} tag`
            }
        }
        const valueType = typeof value === 'boolean'
            ? 'boolean'
            : value.type === 'ExpressionValue'
                ? 'expression'
                : value.type === 'KeyValue'
                    ? 'key'
                    : 'literal'

        const legalTypes = required[key] || optional[key]
        if (!legalTypes.includes(valueType)) {
            return {
                tag: 'Error',
                message: `'${key}' property cannot be of type: ${valueType}`
            }
        }
    }, undefined)
    if (illegalEntry) {
        return illegalEntry
    }
    const missingEntry = Object.keys(required).find((key) => (!(key in open.properties)))
    if (missingEntry) {
        return {
            tag: 'Error',
            message: `'${missingEntry}' property is required on ${open.tag} tag`
        }
    }
    return Object.entries(open.properties)
        .map(([key, value]) => {
            if (typeof value === 'boolean') {
                return [key, value]
            }
            else {
                return [key, value.value]
            }
        })
        .reduce<Record<string, string | boolean>>((previous, [key, value]: [string, string | boolean]) => ({ ...previous, [key]: value }), {}) as T
}

export const isValidateError = (value: Record<string, string | boolean> | ParseError): value is ParseError => (value.tag === 'Error')

export type ExtractProperties<T extends ParseTag, omit extends string> = Omit<T, 'tag' | 'contents' | 'startTagToken' | 'endTagToken' | 'startContentsToken' | 'endContentsToken' | 'closeToken' | omit>
