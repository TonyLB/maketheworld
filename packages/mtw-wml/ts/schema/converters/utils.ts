import { GenericTree } from "../../tree/baseClasses"
import { ParsePropertyTypes, ParseTagOpen, ParseTagSelfClosure } from "../../simpleParser/baseClasses"
import { SchemaTag } from "../baseClasses"
import { ConverterMapValidateProperties, PrintMapOptionsChange, PrintMapOptionsFactory, PrintMode, ValidationTemplate, ValidationTemplateOutput } from "./baseClasses"

export const validateProperties = <V extends ValidationTemplate>(template: V) => (parse: ParseTagOpen | ParseTagSelfClosure): ValidationTemplateOutput<V> => {
    const unmatchedKey = parse.properties.find(({ key }) => (!((key ?? 'DEFAULT') in template)))
    if (unmatchedKey) {
        throw new Error(`Property '${unmatchedKey.key}' is not allowed in '${parse.tag}' items.`)
    }
    const remap = Object.assign({}, ...Object.entries(template).map(([key, { required, type }]) => {
        const matchedKey = parse.properties.find(({ key: checkKey }) => ((checkKey || 'DEFAULT') === key))
        if (required && !matchedKey) {
            throw new Error(`Property '${key}' is required in '${parse.tag}' items.`)
        }
        if (matchedKey && matchedKey.type !== type) {
            const typeLabel = type === ParsePropertyTypes.Boolean ? 'Boolean' : type === ParsePropertyTypes.Expression ? 'Expression' : type === ParsePropertyTypes.Literal ? 'Literal' : 'Key'
            throw new Error(`Property '${key}' must be of ${typeLabel} type in '${parse.tag}' items.`)
        }
        if (matchedKey) {
            return { [key]: matchedKey.value }
        }
        else {
            return {}
        }
    })) as ValidationTemplateOutput<V>
    return remap
}

export const validateContents = ({ isValid, branchTags, leafTags }: ConverterMapValidateProperties) => (contents: GenericTree<SchemaTag>): boolean => {
    return contents.reduce<boolean>((previous, { data: childTag, children }) => {
        if (!previous) {
            return previous
        }
        if (leafTags.includes(childTag.tag)) {
            return isValid(childTag)
        }
        else if (branchTags.includes(childTag.tag)) {
            if (!isValid(childTag)) {
                return false
            }
            else if (children.length) {
                return validateContents({ isValid, branchTags, leafTags })(children)
            }
        }
        return true
    }, true)
}

export const optionsFactory: PrintMapOptionsFactory = (action) => (previous) => {
    switch(action) {
        case PrintMapOptionsChange.Sibling:
            return previous
        case PrintMapOptionsChange.Indent:
            return {
                ...previous,
                forceNest: previous.forceNest === 'breakWrappedLines' ? undefined : previous.forceNest,
                indent: previous.indent + 1
            }
    }
    return previous
}

export const isNaivePrint = (output: string) => (output.split('\n').length <= 1)
export const isNestedPrint = (output: string) => (output.split('\n').length > 1 && Boolean(output.split('\n').find((outputLine) => (outputLine.match(/(?!<:\\)<[^/].*(?!<:\\)>/)))))
export const isPropertyNestedPrint = (output: string) => (output.split('\n').length > 1 && !isNestedPrint(output))

//
// SECTION: Utility functions for dealing with multiple-render listings, and organizing them by nesting level
//

//
// maxIndicesByNestingLevel takes a list of many renders (each with multiple possible levels of nesting and styles of render)
// and lists the maximum indices in each PrintMode
//
export const maxIndicesByNestingLevel = (outputs: string[][]): Record<PrintMode, number> => (
    outputs.reduce<Record<PrintMode, number>>((previous, itemOutputs) => {
        const naiveOutputs = itemOutputs.filter(isNaivePrint)
        const nestedOutputs = itemOutputs.filter(isNestedPrint)
        const propertyNestedOutputs = itemOutputs.filter(isPropertyNestedPrint)
        return {
            [PrintMode.naive]: Math.max(previous[PrintMode.naive], naiveOutputs.length),
            [PrintMode.nested]: Math.max(previous[PrintMode.nested], nestedOutputs.length),
            [PrintMode.propertyNested]: Math.max(previous[PrintMode.propertyNested], propertyNestedOutputs.length),
        }
    }, { [PrintMode.naive]: 0, [PrintMode.nested]: 0, [PrintMode.propertyNested]: 0 })
)

//
// minIndicesByNestingLevel does the same as maxIndicesByNestingLevel, but returns the minimum
//
export const minIndicesByNestingLevel = (outputs: string[][]): Record<PrintMode, number> => (
    outputs.reduce<Record<PrintMode, number>>((previous, itemOutputs) => {
        const naiveOutputs = itemOutputs.filter(isNaivePrint)
        const nestedOutputs = itemOutputs.filter(isNestedPrint)
        const propertyNestedOutputs = itemOutputs.filter(isPropertyNestedPrint)
        return {
            [PrintMode.naive]: Math.min(previous[PrintMode.naive], naiveOutputs.length),
            [PrintMode.nested]: Math.min(previous[PrintMode.nested], nestedOutputs.length),
            [PrintMode.propertyNested]: Math.min(previous[PrintMode.propertyNested], propertyNestedOutputs.length),
        }
    }, { [PrintMode.naive]: Infinity, [PrintMode.nested]: Infinity, [PrintMode.propertyNested]: Infinity })
)

//
// provisionalPrintFactory generates a list of strings selected from the outputs, according to the nestingLevel and index provided.
// So it will search up (for instance) the second Naive interpretation of each among many renders, and return them.
//
export const provisionalPrintFactory = ({ outputs, nestingLevel, indexInLevel }: { outputs: string[][]; nestingLevel: PrintMode; indexInLevel: number }): string[] => {
    const lookup = ({ itemOutputs, nestingLevel, indexInLevel }: { itemOutputs: string[]; nestingLevel: PrintMode; indexInLevel: number }) => {
        if (itemOutputs.length === 0) {
            throw new Error('Empty itemOutput list in provisionalPrintFactory')
        }
        const naiveOutputs = itemOutputs.filter(isNaivePrint)
        const nestedOutputs = itemOutputs.filter(isNestedPrint)
        const propertyNestedOutputs = itemOutputs.filter(isPropertyNestedPrint)
        if (nestingLevel === PrintMode.naive) {
            if (naiveOutputs.length === 0) {
                return lookup({ itemOutputs, nestingLevel: PrintMode.nested, indexInLevel: 0 })
            }
            return naiveOutputs[Math.min(indexInLevel, naiveOutputs.length - 1)]
        }
        else if (nestingLevel === PrintMode.nested) {
            if (nestedOutputs.length === 0) {
                //
                // If there are no nested outputs, default up to naive first, with propertyNested as a fallback
                //
                if (naiveOutputs.length) {
                    return lookup({ itemOutputs, nestingLevel: PrintMode.naive, indexInLevel: naiveOutputs.length - 1 })
                }
                else {
                    return lookup({ itemOutputs, nestingLevel: PrintMode.propertyNested, indexInLevel: 0 })
                }
            }
            return nestedOutputs[Math.min(indexInLevel, nestedOutputs.length - 1)]
        }
        else {
            if (propertyNestedOutputs.length === 0) {
                return lookup({ itemOutputs, nestingLevel: PrintMode.nested, indexInLevel: nestedOutputs.length - 1 })
            }
            return propertyNestedOutputs[Math.min(indexInLevel, propertyNestedOutputs.length - 1)]
        }
    }
    return outputs.map((itemOutputs) => (lookup({ itemOutputs, nestingLevel, indexInLevel })))
}