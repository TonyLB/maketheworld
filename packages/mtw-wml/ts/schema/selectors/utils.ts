import { SchemaTag, isSchemaWithKey } from "../baseClasses";

export const optionsMatch = (options: { tag: string; key: string }) => ({ data }: { data: SchemaTag }): boolean => {
    if (!(options.tag && options.key)) {
        return true
    }
    if (isSchemaWithKey(data) && (data.tag === options.tag) && (data.key === options.key)) {
        return true
    }
    return false
}