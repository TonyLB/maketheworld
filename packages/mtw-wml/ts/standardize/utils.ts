import { excludeUndefined } from "../lib/lists"
import { SchemaTag } from "../schema/baseClasses"
import applyEdits from "../schema/treeManipulation/applyEdits"
import SchemaTagTree from "../tagTree/schema"
import { GenericTreeNode } from "../tree/baseClasses"
import { StandardComponent, StandardNodeKeys } from "./baseClasses"

export const combineTagChildren = <T extends StandardComponent, K extends StandardNodeKeys<T>>(base: T, incoming: T, key: K): T[K] => {
    if (!excludeUndefined(base[key])) {
        return incoming[key]
    }
    if (!excludeUndefined(incoming[key])) {
        return base[key]
    }
    const tagTree = new SchemaTagTree([base[key] as GenericTreeNode<SchemaTag>])
    const incomingTagTree = new SchemaTagTree([incoming[key] as GenericTreeNode<SchemaTag>])
    tagTree._tagList = [...tagTree._tagList, ...incomingTagTree._tagList]
    const combinedSchema = applyEdits(tagTree.tree)
    return { ...combinedSchema[0], id: (base[key] as any).id || (incoming[key] as any).id } as T[K]
}
