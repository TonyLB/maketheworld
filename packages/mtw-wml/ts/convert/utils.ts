import { ParseStackTagOpenEntry, ParseTag, ParseException } from "../parser/baseClasses"
import { parse as acornParse} from "acorn"
import { simple as simpleWalk } from "acorn-walk"
import { SchemaTag } from "../schema/baseClasses"

type ValidatePropertiesValueType = 'boolean' | 'key' | 'expression' | 'literal'

export type ValidatePropertiesItem = Record<string, ValidatePropertiesValueType[]>

type ValidatePropertiesProps = {
    open: ParseStackTagOpenEntry;
    endTagToken: number;
    required: ValidatePropertiesItem;
    optional?: ValidatePropertiesItem;
}

//
// Given a TagOpen token, validateProperties checks against a known set of required and optional properties,
// to syntax check the tokens being parsed.
//
export const validateProperties = <T extends Record<string, any>>({ open, required, endTagToken, optional = {} }: ValidatePropertiesProps): ForceStringType<T> => {
    Object.entries(open.properties).forEach(([key, value]) => {
        if (!(key in required || key in optional)) {
            throw new ParseException(`'${key}' is not a legal property on ${open.tag} tag`, open.startTagToken, endTagToken)
        }
        const valueType = typeof value === 'boolean'
            ? 'boolean'
            : value.type === 'ExpressionValue'
                ? 'expression'
                : value.type === 'KeyValue'
                    ? 'key'
                    : 'literal'

        const legalTypes = required[key] || optional[key]
        if (!legalTypes.includes(valueType)) {
            throw new ParseException(`'${key}' property cannot be of type: ${valueType}`, open.startTagToken, endTagToken)
        }
    })
    const missingEntry = Object.keys(required).find((key) => (!(key in open.properties)))
    if (missingEntry) {
        throw new ParseException(`'${missingEntry}' property is required on ${open.tag} tag`, open.startTagToken, endTagToken)
    }
    return Object.entries(open.properties)
        .map(([key, value]) => {
            if (typeof value === 'boolean') {
                return [key, value]
            }
            else {
                return [key, value.value]
            }
        })
        .reduce<Record<string, any>>((previous, [key, value]: [string, any]) => ({ ...previous, [key]: value }), {}) as ForceStringType<T>
}

export type ExtractProperties<T extends ParseTag, omit extends string> = Omit<T, 'tag' | 'contents' | 'startTagToken' | 'endTagToken' | omit>

export type ForceStringType<T extends Object> = {
    [K in keyof T]: T[K] extends boolean ? boolean : string;
}

type ValidateContentsProps<T extends ParseTag> = {
    contents: ParseTag[];
    legalTags: T["tag"][];
    ignoreTags: ParseTag["tag"][];
}

//
// Given a TagOpen that (at least potentially) has contents, validateContents will confirm that only legal
// child-types are represented in the first level of the content tree.  It will also filter out any tags that
// are passed as needing to be ignored.
//
export const validateContents = <T extends ParseTag>({ contents, legalTags, ignoreTags }: ValidateContentsProps<T>): T[] => {
    return contents.reduce<T[]>((previous, contentItem) => {
        if (legalTags.includes(contentItem.tag)) {
            return [...previous, contentItem as T]
        }
        if (!ignoreTags.includes(contentItem.tag)) {
            throw new ParseException(`'${contentItem.tag}' tag not allowed in this context`, contentItem.startTagToken, contentItem.endTagToken)
        }
        return previous
    }, [])
}

//
// extractDependenciesFromJS is a painfully naive dependency extractor using only the barest fraction of the recursive
// scoping functionality of acorn-walk ... and will still probably be good enough for 99+% of cases
//
export const extractDependenciesFromJS = (src: string): string[] => {
    const parsedJS = acornParse(src.trim(), { ecmaVersion: 'latest' })
    let identifiedGlobals: string[] = []
    let definedLocals: string[] = []
    simpleWalk(parsedJS, {
        Identifier(node) {
            const identifier = (node as any).name
            if (!(definedLocals.includes(identifier))) {
                identifiedGlobals.push(identifier)
            }
        },
        ArrowFunctionExpression(node) {
            ((node as any).params || []).forEach(({ name }) => {
                definedLocals.push(name)
            })
        },
        VariableDeclarator(node) {
            definedLocals.push((node as any).id.name)
        }
    })
    return [...(new Set(identifiedGlobals.filter((item) => (!definedLocals.includes(item)))))]
}

export const indentSpacing = (indent: number): string => {
    return '    '.repeat(indent)
}

export const lineLengthAfterIndent = (indent: number): number => (Math.max(40, 80 - indent * 4))

type TagRenderProperty = {
    key: string;
    type: 'key' | 'expression' | 'literal';
    value: string;
} | {
    key: string;
    type: 'boolean';
    value: boolean;
}

export const tagRender = ({ indent, forceNest, tag, properties, contents }: { indent: number, forceNest?: boolean, tag: string, properties: TagRenderProperty[]; contents: string[]; }): string => {
    const propertyRender = properties.map((property) => {
        switch(property.type) {
            case 'boolean':
                return property.value ? `${property.key}` : ''
            case 'expression':
                return property.value ? `${property.key}={${property.value}}` : ''
            case 'key':
                return property.value ? `${property.key}=(${property.value})` : ''
            case 'literal':
                return property.value ? `${property.key}="${property.value}"` : ''
        }
    }).filter((value) => (value))
    const tagOpen = `<${[tag, ...propertyRender].join(' ')}>`
    const tagClose = `</${tag}>`
    const naive = `${tagOpen}${contents.join('')}${tagClose}`
    const nested = `${[tagOpen, ...contents].join(`\n${indentSpacing(indent + 1)}`)}\n${indentSpacing(indent)}${tagClose}`
    if (typeof forceNest === 'undefined') {
        return (naive.length > lineLengthAfterIndent(indent)) ? nested : naive
    }
    else {
        return forceNest ? nested : naive
    }
}

export const makeSchemaTag = <T extends Omit<SchemaTag, 'parse'>>(tag: T): SchemaTag => {
    return {
        ...tag,
        parse: {
            tag: 'Space',
            startTagToken: 0,
            endTagToken: 0
        }
    } as SchemaTag
}
