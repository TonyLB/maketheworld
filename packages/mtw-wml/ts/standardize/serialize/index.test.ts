import { Schema } from '../../schema'
import { Standardizer } from '..'
import { deIndentWML } from '../../schema/utils'
import { serialize } from '.'

describe('standard form serialize' ,() => {
    it('should return single line on empty asset', () => {
        expect(serialize({
            tag: 'Asset',
            key: 'Test',
            byId: {},
            metaData: []
        })).toEqual([{ tag: 'Asset', key: 'Test' }])
    })

    it('should serialize all component types', () => {
        const schema = new Schema()
        schema.loadWML(deIndentWML(`
            <Asset key=(test)>
                <Image key=(testIcon) />
                <Variable key=(open) default={false} />
                <Action key=(toggleOpen) src={open = !open} />
                <Computed key=(closed) src={!open} />
                <Room key=(testRoom)>
                    <ShortName>Vortex</ShortName>
                    <Name>Vortex</Name>
                    <Description>Vortex Desc</Description>
                </Room>
                <Feature key=(testFeature)>
                    <Name>Clocktower</Name>
                    <Description>A tower built of white sandstone blocks, with an ornate clock set on the northern face.</Description>
                </Feature>
                <Knowledge key=(testKnowledge)>
                    <Name>Learn</Name>
                    <Description>
                        There is so much to know!
                    </Description>
                </Knowledge>
                <Moment key=(openDoorMoment)>
                    <Message key=(openDoor)>
                        The door opens!
                        <Room key=(testRoom) />
                    </Message>
                </Moment>
            </Asset>
        `))
        const standard = new Standardizer(schema.schema)
        expect(serialize(standard.stripped)).toMatchSnapshot()
    })

    it('should serialize imports', () => {
        const schema = new Schema()
        schema.loadWML(deIndentWML(`
            <Asset key=(test)>
                <Import from=(testImport)><Room key=(testIn) as=(testRoom) /></Import>
                <Room key=(testRoom)>
                    <ShortName>Test</ShortName>
                </Room>
            </Asset>
        `))
        const standard = new Standardizer(schema.schema)
        expect(serialize(standard.stripped)).toEqual([
            { tag: 'Asset', key: 'test' },
            {
                tag: 'Room',
                from: { assetId: 'testImport', key: 'testIn' },
                key: 'testRoom',
                shortName: { data: { tag: 'ShortName' }, children: [{ data: { tag: 'String', value: 'Test' }, children: [] }] },
                name: { data: { tag: 'Name' }, children: [] },
                summary: { data: { tag: 'Summary' }, children: [] },
                description: { data: { tag: 'Description' }, children: [] },
                exits: [],
                themes: []
            }
        ])
    })

    it('should serialize exports', () => {
        const schema = new Schema()
        schema.loadWML(deIndentWML(`
            <Asset key=(test)>
                <Room key=(testRoom)>
                    <ShortName>Test</ShortName>
                </Room>
                <Export><Room key=(testRoom) as=(Room3) /></Export>
            </Asset>
        `))
        const standard = new Standardizer(schema.schema)
        expect(serialize(standard.stripped)).toEqual([
            { tag: 'Asset', key: 'test' },
            {
                tag: 'Room',
                key: 'testRoom',
                exportAs: 'Room3',
                shortName: { data: { tag: 'ShortName' }, children: [{ data: { tag: 'String', value: 'Test' }, children: [] }] },
                name: { data: { tag: 'Name' }, children: [] },
                summary: { data: { tag: 'Summary' }, children: [] },
                description: { data: { tag: 'Description' }, children: [] },
                exits: [],
                themes: []
            }
        ])
    })

})