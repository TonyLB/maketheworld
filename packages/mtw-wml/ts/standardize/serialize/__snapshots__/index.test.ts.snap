// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`standard form serialize should serialize a character 1`] = `
Array [
  Object {
    "fileName": undefined,
    "firstImpression": Object {
      "children": Array [],
      "data": Object {
        "tag": "FirstImpression",
        "value": "Frumpy Goth",
      },
    },
    "image": Object {
      "children": Array [],
      "data": Object {
        "key": "TessIcon",
        "tag": "Image",
      },
    },
    "key": "Tess",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Tess",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
    },
    "oneCoolThing": Object {
      "children": Array [],
      "data": Object {
        "tag": "OneCoolThing",
        "value": "Fuchsia eyes",
      },
    },
    "outfit": Object {
      "children": Array [],
      "data": Object {
        "tag": "Outfit",
        "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
      },
    },
    "pronouns": Object {
      "children": Array [],
      "data": Object {
        "adjective": "hers",
        "object": "her",
        "possessive": "her",
        "reflexive": "herself",
        "subject": "she",
        "tag": "Pronouns",
      },
    },
    "tag": "Character",
    "universalKey": "CHARACTER#001",
  },
  Object {
    "fileName": "IMAGE-002",
    "key": "TessIcon",
    "tag": "Image",
    "universalKey": "IMAGE#002",
  },
]
`;

exports[`standard form serialize should serialize all component types 1`] = `
Array [
  Object {
    "key": "test",
    "tag": "Asset",
    "universalId": "ASSET#test",
  },
  Object {
    "fileName": "IMAGE-001",
    "key": "testBackground",
    "tag": "Image",
    "universalKey": "IMAGE#001",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex Desc",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
      "wrapperTag": undefined,
    },
    "exits": Array [],
    "fileName": undefined,
    "key": "testRoom",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
      "wrapperTag": undefined,
    },
    "shortName": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex",
          },
        },
      ],
      "data": Object {
        "tag": "ShortName",
      },
      "wrapperTag": undefined,
    },
    "summary": Object {
      "children": Array [],
      "data": Object {
        "tag": "Summary",
      },
    },
    "tag": "Room",
    "themes": Array [],
    "universalKey": "ROOM#002",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "A tower built of white sandstone blocks, with an ornate clock set on the northern face.",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
      "wrapperTag": undefined,
    },
    "fileName": undefined,
    "key": "testFeature",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Clocktower",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
      "wrapperTag": undefined,
    },
    "tag": "Feature",
    "universalKey": "FEATURE#003",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "There is so much to know!",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
      "wrapperTag": undefined,
    },
    "fileName": undefined,
    "key": "testKnowledge",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Learn",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
      "wrapperTag": undefined,
    },
    "tag": "Knowledge",
    "universalKey": "KNOWLEDGE#004",
  },
  Object {
    "fileName": undefined,
    "images": Array [
      Object {
        "children": Array [],
        "data": Object {
          "key": "testBackground",
          "tag": "Image",
        },
      },
    ],
    "key": "testMap",
    "name": Object {
      "children": Array [],
      "data": Object {
        "tag": "Name",
      },
    },
    "positions": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "data": Object {
              "tag": "Position",
              "x": 0,
              "y": 100,
            },
          },
        ],
        "data": Object {
          "key": "testRoom",
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
        "wrapperTag": undefined,
      },
    ],
    "tag": "Map",
    "themes": Array [],
    "universalKey": "MAP#005",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "The door opens!",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
    },
    "fileName": undefined,
    "key": "openDoor",
    "rooms": Array [
      Object {
        "children": Array [],
        "data": Object {
          "key": "testRoom",
          "tag": "Room",
          "x": undefined,
          "y": undefined,
        },
      },
    ],
    "tag": "Message",
    "universalKey": "MESSAGE#006",
  },
  Object {
    "fileName": undefined,
    "key": "openDoorMoment",
    "messages": Array [
      Object {
        "children": Array [],
        "data": Object {
          "key": "openDoor",
          "tag": "Message",
        },
      },
    ],
    "tag": "Moment",
    "universalKey": "MOMENT#007",
  },
  Object {
    "default": "false",
    "fileName": undefined,
    "key": "open",
    "tag": "Variable",
    "universalKey": "VARIABLE#008",
  },
  Object {
    "fileName": undefined,
    "key": "closed",
    "src": "!open",
    "tag": "Computed",
    "universalKey": "COMPUTED#009",
  },
  Object {
    "fileName": undefined,
    "key": "toggleOpen",
    "src": "open = !open",
    "tag": "Action",
    "universalKey": "ACTION#010",
  },
]
`;

exports[`standard form serialize should serialize universal keys 1`] = `
Array [
  Object {
    "key": "test",
    "tag": "Asset",
    "universalId": "ASSET#test",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex Desc",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
      "wrapperTag": undefined,
    },
    "exits": Array [],
    "fileName": undefined,
    "key": "testRoom",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
      "wrapperTag": undefined,
    },
    "shortName": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Vortex",
          },
        },
      ],
      "data": Object {
        "tag": "ShortName",
      },
      "wrapperTag": undefined,
    },
    "summary": Object {
      "children": Array [],
      "data": Object {
        "tag": "Summary",
      },
    },
    "tag": "Room",
    "themes": Array [],
    "universalKey": "ROOM#UUID",
  },
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "A tower built of white sandstone blocks, with an ornate clock set on the northern face.",
          },
        },
      ],
      "data": Object {
        "tag": "Description",
      },
      "wrapperTag": undefined,
    },
    "fileName": undefined,
    "key": "testFeature",
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "tag": "String",
            "value": "Clocktower",
          },
        },
      ],
      "data": Object {
        "tag": "Name",
      },
      "wrapperTag": undefined,
    },
    "tag": "Feature",
    "universalKey": "FEATURE#UUID",
  },
]
`;
